schnorr_fun/src/frost/chilldkg/encpedpop.rs

```diff
@@ -216,82 +216,45 @@ impl AggKeygenInput {
-        let mut hash_state = H::default();
-        if !fingerprint.tag.is_empty() {
-            hash_state = hash_state.add([fingerprint.tag.len() as u8]);
-            hash_state = hash_state.add(fingerprint.tag.as_bytes());
-        }
-        hash_state = hash_state.add(public_key);
-
-        let mut tweaks = Vec::with_capacity(self.inner.agg_poly.len());
-
-        // Grind each coefficient in sequence, note that agg_poly only
-        // contains the non-constant coefficients.
-        for coeff_index in 0..self.inner.agg_poly.len() {
-            let mut total_tweak = Scalar::<Public, Zero>::zero();
-            let original_coeff = self.inner.agg_poly[coeff_index];
-            let mut current_coeff = original_coeff;
-
-            loop {
-                // Clone the hash state from previous coefficients and add current one
-                let hash = hash_state.clone().add(current_coeff);
-                // Check if hash has required number of leading zero bits
-                let hash_bytes: [u8; 32] = hash.clone().finalize_fixed().into();
-                if Fingerprint::leading_zero_bits(&hash_bytes[..])
-                    >= fingerprint.bit_length as usize
-                {
-                    // Update hash_state for next coefficient because the next coefficient hash
-                    // includes the current one.
-                    hash_state = hash;
-                    break;
-                }
-
-                // Add one more G to the coefficient
-                total_tweak += s!(1);
-                current_coeff = g!(current_coeff + G).normalize();
-            }
-
-            // Apply the final tweak to the polynomial coefficient
-            self.inner.agg_poly[coeff_index] = current_coeff;
-            tweaks.push(total_tweak);
-        }
-
-        // Apply all tweaks to encrypted shares at once
-        for (share_index, (_enc_key, encrypted_share)) in &mut self.encrypted_shares {
-            let mut power = *share_index;
-            for tweak in &tweaks {
-                *encrypted_share += s!(tweak * power).public();
-                power *= share_index;
-            }
+        let mut shared_key = self.shared_key();
+        let tweaks = shared_key.grind_fingerprint::<H>(fingerprint);
+        // replace our poly with the one that has the fingerprint
+        self.inner.agg_poly = shared_key.point_polynomial()[1..].to_vec();
+        debug_assert!(self.shared_key().check_fingerprint::<H>(&fingerprint));
+
+        for (share_index, (_encryption_key, encrypted_secret_share)) in &mut self.encrypted_shares {
+            // 💡 The share encryption is homomorphic so we can apply the tweak
+            // operations as if the coordinator had local copies of the secret
+            // shares.
+            let mut tmp = SecretShare {
+                index: *share_index,
+                share: encrypted_secret_share.secret(/*HACK: we need a secret one to use the function*/),
```

Comment:
`encrypted_secret_share.secret()` was quite confusing since it sounds like we're getting the secret out of the encrypted secret share, not *marking* it as secret for the API.

Checking understanding:

We find grind some tweaks to the non-constant terms of the polynomial.
Then we overwrite the non-constant terms of `agg_poly` from the grinded polynomial.

Then finally, we modify all the encrypted secret share by adding the term to have it sit on this tweaked polynomial?

Reply:
yes. I made it easier to read by having `SecretShare` take a type parameter for the secrecy so you don't have to cast it.